From caf99ed70dd729790530bacf3f473e4d9357edb4 Mon Sep 17 00:00:00 2001
From: Yurii Rashkovskii <me@yrashk.com>
Date: Mon, 7 May 2018 13:17:17 -0700
Subject: [PATCH] Problem: Repository::new_named_item duplicate check on
 aarch64

Tests pertaining to duplicate name fail in debug mode on aarch64,
the OS error is never returned.

Solution: work around the problem specifically for this environment

This patch will ensure that the existence of the directory is
checked on aarch64 in debug mode. It is not perfect, however,
because there's still a gap between the check and the attempt to
create the directory (however small it might be to a human eye)

Tracking issue has been filed with Rust:
https://github.com/rust-lang/rust/issues/50516
---
 .../2UT54I7HQDVAOTKROJO5XHKPAETRWMQF/.authors                      | 1 +
 .../2UT54I7HQDVAOTKROJO5XHKPAETRWMQF/.timestamp                    | 1 +
 .../2UT54I7HQDVAOTKROJO5XHKPAETRWMQF/.type/SummaryChanged          | 0
 .../2UT54I7HQDVAOTKROJO5XHKPAETRWMQF/text                          | 1 +
 .../JESGGRQLIKCQCF4OKGNVXQSMMWLDBHQW/.authors                      | 1 +
 .../.prev/2UT54I7HQDVAOTKROJO5XHKPAETRWMQF                         | 0
 .../JESGGRQLIKCQCF4OKGNVXQSMMWLDBHQW/.timestamp                    | 1 +
 .../JESGGRQLIKCQCF4OKGNVXQSMMWLDBHQW/.type/DetailsChanged          | 0
 .../JESGGRQLIKCQCF4OKGNVXQSMMWLDBHQW/text                          | 5 +++++
 sit-core/src/repository.rs                                         | 7 +++++++
 10 files changed, 17 insertions(+)
 create mode 100644 .sit/items/ed4bbb0c-2973-4de1-b5f8-648d3011ae86/2UT54I7HQDVAOTKROJO5XHKPAETRWMQF/.authors
 create mode 100644 .sit/items/ed4bbb0c-2973-4de1-b5f8-648d3011ae86/2UT54I7HQDVAOTKROJO5XHKPAETRWMQF/.timestamp
 create mode 100644 .sit/items/ed4bbb0c-2973-4de1-b5f8-648d3011ae86/2UT54I7HQDVAOTKROJO5XHKPAETRWMQF/.type/SummaryChanged
 create mode 100644 .sit/items/ed4bbb0c-2973-4de1-b5f8-648d3011ae86/2UT54I7HQDVAOTKROJO5XHKPAETRWMQF/text
 create mode 100644 .sit/items/ed4bbb0c-2973-4de1-b5f8-648d3011ae86/JESGGRQLIKCQCF4OKGNVXQSMMWLDBHQW/.authors
 create mode 100644 .sit/items/ed4bbb0c-2973-4de1-b5f8-648d3011ae86/JESGGRQLIKCQCF4OKGNVXQSMMWLDBHQW/.prev/2UT54I7HQDVAOTKROJO5XHKPAETRWMQF
 create mode 100644 .sit/items/ed4bbb0c-2973-4de1-b5f8-648d3011ae86/JESGGRQLIKCQCF4OKGNVXQSMMWLDBHQW/.timestamp
 create mode 100644 .sit/items/ed4bbb0c-2973-4de1-b5f8-648d3011ae86/JESGGRQLIKCQCF4OKGNVXQSMMWLDBHQW/.type/DetailsChanged
 create mode 100644 .sit/items/ed4bbb0c-2973-4de1-b5f8-648d3011ae86/JESGGRQLIKCQCF4OKGNVXQSMMWLDBHQW/text

diff --git a/.sit/items/ed4bbb0c-2973-4de1-b5f8-648d3011ae86/2UT54I7HQDVAOTKROJO5XHKPAETRWMQF/.authors b/.sit/items/ed4bbb0c-2973-4de1-b5f8-648d3011ae86/2UT54I7HQDVAOTKROJO5XHKPAETRWMQF/.authors
new file mode 100644
index 0000000..8022052
--- /dev/null
+++ b/.sit/items/ed4bbb0c-2973-4de1-b5f8-648d3011ae86/2UT54I7HQDVAOTKROJO5XHKPAETRWMQF/.authors
@@ -0,0 +1 @@
+Yurii Rashkovskii <me@yrashk.com>
\ No newline at end of file
diff --git a/.sit/items/ed4bbb0c-2973-4de1-b5f8-648d3011ae86/2UT54I7HQDVAOTKROJO5XHKPAETRWMQF/.timestamp b/.sit/items/ed4bbb0c-2973-4de1-b5f8-648d3011ae86/2UT54I7HQDVAOTKROJO5XHKPAETRWMQF/.timestamp
new file mode 100644
index 0000000..b691f4f
--- /dev/null
+++ b/.sit/items/ed4bbb0c-2973-4de1-b5f8-648d3011ae86/2UT54I7HQDVAOTKROJO5XHKPAETRWMQF/.timestamp
@@ -0,0 +1 @@
+2018-05-07T20:16:50.974Z
\ No newline at end of file
diff --git a/.sit/items/ed4bbb0c-2973-4de1-b5f8-648d3011ae86/2UT54I7HQDVAOTKROJO5XHKPAETRWMQF/.type/SummaryChanged b/.sit/items/ed4bbb0c-2973-4de1-b5f8-648d3011ae86/2UT54I7HQDVAOTKROJO5XHKPAETRWMQF/.type/SummaryChanged
new file mode 100644
index 0000000..e69de29
diff --git a/.sit/items/ed4bbb0c-2973-4de1-b5f8-648d3011ae86/2UT54I7HQDVAOTKROJO5XHKPAETRWMQF/text b/.sit/items/ed4bbb0c-2973-4de1-b5f8-648d3011ae86/2UT54I7HQDVAOTKROJO5XHKPAETRWMQF/text
new file mode 100644
index 0000000..e00d738
--- /dev/null
+++ b/.sit/items/ed4bbb0c-2973-4de1-b5f8-648d3011ae86/2UT54I7HQDVAOTKROJO5XHKPAETRWMQF/text
@@ -0,0 +1 @@
+Problem: workaround for std::fs:create_dir not failing on existing directory on aarch64, debug
\ No newline at end of file
diff --git a/.sit/items/ed4bbb0c-2973-4de1-b5f8-648d3011ae86/JESGGRQLIKCQCF4OKGNVXQSMMWLDBHQW/.authors b/.sit/items/ed4bbb0c-2973-4de1-b5f8-648d3011ae86/JESGGRQLIKCQCF4OKGNVXQSMMWLDBHQW/.authors
new file mode 100644
index 0000000..8022052
--- /dev/null
+++ b/.sit/items/ed4bbb0c-2973-4de1-b5f8-648d3011ae86/JESGGRQLIKCQCF4OKGNVXQSMMWLDBHQW/.authors
@@ -0,0 +1 @@
+Yurii Rashkovskii <me@yrashk.com>
\ No newline at end of file
diff --git a/.sit/items/ed4bbb0c-2973-4de1-b5f8-648d3011ae86/JESGGRQLIKCQCF4OKGNVXQSMMWLDBHQW/.prev/2UT54I7HQDVAOTKROJO5XHKPAETRWMQF b/.sit/items/ed4bbb0c-2973-4de1-b5f8-648d3011ae86/JESGGRQLIKCQCF4OKGNVXQSMMWLDBHQW/.prev/2UT54I7HQDVAOTKROJO5XHKPAETRWMQF
new file mode 100644
index 0000000..e69de29
diff --git a/.sit/items/ed4bbb0c-2973-4de1-b5f8-648d3011ae86/JESGGRQLIKCQCF4OKGNVXQSMMWLDBHQW/.timestamp b/.sit/items/ed4bbb0c-2973-4de1-b5f8-648d3011ae86/JESGGRQLIKCQCF4OKGNVXQSMMWLDBHQW/.timestamp
new file mode 100644
index 0000000..b691f4f
--- /dev/null
+++ b/.sit/items/ed4bbb0c-2973-4de1-b5f8-648d3011ae86/JESGGRQLIKCQCF4OKGNVXQSMMWLDBHQW/.timestamp
@@ -0,0 +1 @@
+2018-05-07T20:16:50.974Z
\ No newline at end of file
diff --git a/.sit/items/ed4bbb0c-2973-4de1-b5f8-648d3011ae86/JESGGRQLIKCQCF4OKGNVXQSMMWLDBHQW/.type/DetailsChanged b/.sit/items/ed4bbb0c-2973-4de1-b5f8-648d3011ae86/JESGGRQLIKCQCF4OKGNVXQSMMWLDBHQW/.type/DetailsChanged
new file mode 100644
index 0000000..e69de29
diff --git a/.sit/items/ed4bbb0c-2973-4de1-b5f8-648d3011ae86/JESGGRQLIKCQCF4OKGNVXQSMMWLDBHQW/text b/.sit/items/ed4bbb0c-2973-4de1-b5f8-648d3011ae86/JESGGRQLIKCQCF4OKGNVXQSMMWLDBHQW/text
new file mode 100644
index 0000000..b354cbc
--- /dev/null
+++ b/.sit/items/ed4bbb0c-2973-4de1-b5f8-648d3011ae86/JESGGRQLIKCQCF4OKGNVXQSMMWLDBHQW/text
@@ -0,0 +1,5 @@
+std::fs::create_dir does not fail on existing directory on aarch64, when built in debug, leading to failing tests and change in behavior (item with the same name can be created more than once). This issue has been resolved with a workaround handling this special case specifically in this configuration.
+
+However, this is not the most beautiful solution.
+
+Tracking issue filed with Rust: https://github.com/rust-lang/rust/issues/50516
\ No newline at end of file
diff --git a/sit-core/src/repository.rs b/sit-core/src/repository.rs
index ebd357d..501ea33 100644
--- a/sit-core/src/repository.rs
+++ b/sit-core/src/repository.rs
@@ -305,6 +305,13 @@ impl Repository {
         let id: String = name.into();
         let mut path = self.items_path.clone();
         path.push(&id);
+        // https://github.com/rust-lang/rust/issues/50516
+        #[cfg(all(debug_assertions, target_arch = "aarch64"))] {
+          use std::io;
+          if path.is_dir() {
+             return Err(io::Error::from_raw_os_error(17).into()); // 17 is EEXIST
+          }
+        }
         fs::create_dir(path)?;
         let id = OsString::from(id);
         Ok(Item {
-- 
2.16.2

